backends/p4tools/testgen/targets/bmv2/expr_stepper.cpp

```diff
@@ -0,0 +1,1742 @@
+             auto* nextState = new ExecutionState(state);
+             nextState->popBody();
+             result->emplace_back(nextState);
+         }},
+        /* ======================================================================================
+         *  clone_preserving_field_list
+         *  Calling clone_preserving_field_list during execution of the ingress
+         *  or egress control will cause the packet to be cloned, sometimes
+         *  also called mirroring, i.e. zero or more copies of the packet are
+         *  made, and each will later begin egress processing as an independent
+         *  packet from the original packet.  The original packet continues
+         *  with its normal next steps independent of the clone(s).
+         *
+         *  The session parameter is an integer identifying a clone session id
+         *  (sometimes called a mirror session id).  The control plane software
+         *  must configure each session you wish to use, or else no clones will
+         *  be made using that session.  Typically this will involve the
+         *  control plane software specifying one output port to which the
+         *  cloned packet should be sent, or a list of (port, egress_rid) pairs
+         *  to which a separate clone should be created for each, similar to
+         *  multicast packets.
+         *
+         *  Cloned packets can be distinguished from others by the value of the
+         *  standard_metadata instance_type field.
+         *
+         *  The user metadata fields that are tagged with @field_list(index) will be
+         *  sent to the parser together with a clone of the packet.
+         *
+         *  If clone_preserving_field_list is called during ingress processing,
+         *  the first parameter must be CloneType.I2E.  If
+         *  clone_preserving_field_list is called during egress processing, the
+         *  first parameter must be CloneType.E2E.
+         *
+         *  There is no way to undo its effects once it has been called.  If
+         *  there are multiple calls to clone_preserving_field_list and/or
+         *  clone during a single execution of the same ingress (or egress)
+         *  control, only the last clone session and index are used.  See the
+         *  v1model architecture documentation (Note 1) for more details.
+         * ====================================================================================== */
+        {"*method.clone_preserving_field_list",
+         {"type", "session", "data"},
+         [this](const IR::MethodCallExpression* call, const IR::Expression* /*receiver*/,
+                IR::ID& /*methodName*/, const IR::Vector<IR::Argument>* args,
+                const ExecutionState& state, SmallStepEvaluator::Result& result) {
+             uint64_t recirculateCount = 0;
+             // Grab the recirculate count. Stop after more than 1 circulation loop to avoid
+             // infinite recirculation loops.
```

Comment:
Is this part of the bmv2 spec?

Reply:
No, in fact BMv2 seems to circulate more rounds. I do not have the exact count. Will add this as a TODO.

Reply:
I am pretty sure that bmv2 does not enforce any limit on the number of recirculations or clones of a packet.